src/components/tl/mlx5/mcast/tl_mlx5_mcast_rcache.c

```diff
@@ -0,0 +1,161 @@
+/**
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_mlx5_mcast_rcache.h"
+
+static ucc_status_t ucc_tl_mlx5_mcast_coll_reg_mr(ucc_tl_mlx5_mcast_coll_context_t *ctx,
+                                                  void *data, size_t data_size, void **mr)
+{
+    *mr = ibv_reg_mr(ctx->pd, data, data_size, IBV_ACCESS_LOCAL_WRITE,
+                     IBV_ACCESS_REMOTE_READ, IBV_ACCESS_REMOTE_WRITE);
+
+    tl_debug(ctx->lib, "External memory register: ptr %p, len %zd, mr %p",
+            data,  data_size, (*mr));
+    if (!*mr) {
+        tl_error(ctx->lib, "TL/MCAST: Failed to register MR\n");
+        return UCC_ERR_NO_MEMORY;
+    }
+    
+    return UCC_OK;
+}
+
+
+static ucc_status_t ucc_tl_mlx5_mcast_coll_dereg_mr(ucc_tl_mlx5_mcast_coll_context_t *ctx, void *mr)
+{
+    if(mr != NULL) {
```

Comment:
Can you explain why you need this check ?
Maybe I'm guessing wrong, but if we keep this check we might also want to set mr to NULL after the deregistration, so we can safely call `ucc_tl_mlx5_mcast_coll_dereg_mr` again

Reply:
Please note that this is a self-contained rcache library and we must be prepared for anything that users of these API inside rcache.h throw at us. That is why we must check input mr to make sure it is not NULL. Setting MR to NULL is the responsibility of the user who calls dereg function.

Reply:
please rewrite it without else e.g.
if (ucc_unlikely(mr == NULL)) {
    tl_debug(....)
    return UCC_OK;
}

everything else